src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -0,0 +1,627 @@
+//
+// NSUrlSessionHandler.cs:
+//
+// Authors:
+//     Paul Betts <paul@paulbetts.org>
+//     Nick Berardi <nick@nickberardi.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+#if UNIFIED
+using CoreFoundation;
+using Foundation;
+using Security;
+#else
+using MonoTouch.CoreFoundation;
+using MonoTouch.Foundation;
+using MonoTouch.Security;
+using System.Globalization;
+using nint = System.Int32;
+using nuint = System.UInt32;
+#endif
+
+#if SYSTEM_NET_HTTP
+namespace System.Net.Http {
+#else
+namespace Foundation {
+#endif
+	public partial class NSUrlSessionHandler : HttpMessageHandler
+	{
+		private readonly Dictionary<string, string> headerSeparators = new Dictionary<string, string> {
+			["User-Agent"] = " ",
+			["Server"] = " "
+		};
+
+		private readonly NSUrlSession session;
+		private readonly Dictionary<NSUrlSessionTask, InflightData> inflightRequests;
+		private readonly object inflightRequestsLock = new object ();
+
+		public NSUrlSessionHandler ()
+		{
+			var configuration = NSUrlSessionConfiguration.DefaultSessionConfiguration;
+
+			// we cannot do a bitmask but we can set the minimum based on ServicePointManager.SecurityProtocol minimum
+			var sp = ServicePointManager.SecurityProtocol;
+			if ((sp & SecurityProtocolType.Ssl3) != 0)
```

Comment:
This defaults to the least secure option (Ssl3), maybe we should reverse the conditions here to put TLS 1.2 first?